cfme/v2v/migrations.py

```diff
@@ -125,10 +125,12 @@ def fill(self, values):
                        ...
                     }
         """
-        for mapping in values['mappings']:
-            self.source_clusters.fill(mapping['sources'])
-            self.target_clusters.fill(mapping['target'])
-            self.add_mapping.click()
+        if values:
```

Comment:
:+1:

Reply:
```In[10]: a = {'mappings':[]}
In[11]: if a:
   ...:     print "This is a test"
   ...: 
This is a test
```
I don't think this test really gets you very much. We should explicitly check that all the right keys are there.

Reply:
@jawatts : These Views are filled only if the form_data has respective data in it .
For ex: If 
form_data = {
        'general': {
            'description': "my edited description"},
        'cluster': {},
        'datastore': {},
        'network': {}
    }

So if cluster is present only then that view is filled . and if values check if it has mappings or empty , so working in this case

Reply:
I think you two are in a way talking across each other here, both in a way missing the point a bit :)

1. Justin's correct, checking values is empty before iterating without actually doing something with that information (like logging that the clusters fill was empty) doesn't get us much.

2. Shveta's correct too, its not a problem that its empty in the fill context, and we just ultimately want to click next after skipping this.

All of this is due to the fact that we have a fill dictionary (that really should be a list anyway) that doesn't match our view/widget structure and names, so we have to try and mangle the data to fill, instead of letting `View.fill` manage it for us.

In this case, I don't think you should add a check of `if values`, because `values['mappings']` could still raise an exception anyway (the check is incomplete).

Simply using `for mapping in (values.get('mappings', []) if isinstance(values, dict) else []):` protects from the KeyError, values not being a dict, and short-circuits the loop if the `values` dict is empty or contains an unexpected key.

I have plenty of comments about the fill dicts not matching the view structure, but I think that might be out of scope for the changes proposed here.  Doing so would remove our need to have these corner cases during fill in the first place.

```
In [3]: values = {}                                                                                                                        

In [4]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
   ...:     print('true') 
   ...:                                                                                                                                    

In [5]: values = {'thing'}                                                                                                                 

In [6]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
   ...:     print('true') 
   ...:                                                                                                                                    

In [7]: values = []                                                                                                                        

In [8]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
   ...:     print('true') 
   ...:                                                                                                                                    

In [9]: values = 'stick'                                                                                                                   

In [10]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
    ...:     print('true') 
    ...:                                                                                                                                   

In [11]: values = {'mappings':[]}                                                                                                          

In [12]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
    ...:     print('true') 
    ...:      
    ...:                                                                                                                                   

In [13]: values = {'mappings':[1]}                                                                                                         

In [14]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
    ...:     print('true') 
    ...:      
    ...:                                                                                                                                   
true

In [15]: values = {'mappings':[1, 2]}                                                                                                      

In [16]: for mapping in (values.get('mappings', []) if isinstance(values, dict) else []): 
    ...:     print('true') 
    ...:      
    ...:                                                                                                                                   
true
true

```